File: Algorithms/Two-Pointers/0202_happy_number.py
#two_pointers

class Solution(object):
    def isHappy(self, n):
        """
        :type n: int
        :rtype: bool
        """
        # positive, replace num with sum of square of digits until = 1

        def next_num(n):
            output = 0

            while n: 
                digit = n % 10 
                output += digit ** 2
                n = n // 10

            return output
        
        slow = next_num(n)
        #move fast ptr twice as fast 
        fast = next_num(next_num(n))

        while slow != fast: 
            if fast == 1: return True 
            slow = next_num(slow)
            fast = next_num(next_num(fast))
        
        return slow == 1

File: Algorithms/Two-Pointers/0202_happy_number.c
// Practice for: 0202_happy_number.py
// Bridge logic from Python to C pointers here.

#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h> 

//floyd's cycle detection algo (tortise and hare algo)

int nextNum (int n){
    int output = 0; 

    while (n != 0){
        int digit = n % 10; 
        //remember that c has no ** function
        output += digit * digit;
        n /= 10;
    }
    return output;
}

bool isHappy(int n){
    int slow = nextNum(n);
    int fast = nextNum(nextNum(n));

    while (slow != fast){
        if (fast == 1){
            return true;
        }
        slow = nextNum(slow);
        fast = nextNum(nextNum(fast));

    }
    return slow == 1;
}
